### PR TITLE
KAFKA-5062. Kafka brokers can accept malformed requests which allocat…

### PR DESCRIPTION
**Pull request by cmccabe**
…e gigabytes of memory

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5419/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5404/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by edoardocomar** [comment]
@cmccabe - we have reproduced this issue but not published its results on the wide mailing list as it makes a DOS attack on Kafka feasible.
Happy to discuss separately 
cc @rajinisivaram @ijuma @mimaison

**Comment by junrao** [comment]
Just a quick question. Does the patch actually tighten up any of the validation or is it just a better implementation of the existing logic? From a quick look at the code, it seems to me the patch is the latter?

**Comment by ijuma** [comment]
Yes, I was wondering the same. We check the api keys, etc. in `RequestChannel` today. Also, it would be good to include tests if this is indeed fixing a bug.

**Comment by cmccabe** [comment]
> Just a quick question. Does the patch actually tighten up any of the validation or is it just a better implementation of the existing logic? From a quick look at the code, it seems to me the patch is the latter?

The validation is exactly the same, but we refrain from allocating a buffer until the api key and api version have been verified.

File: clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java [hunk 1]
```diff
@@ -26,35 +28,165 @@
...
+            ApiKeys api;
+            apiVersion = startBuffer.getShort();
+            try {
+                api = ApiKeys.forId(apiKey);
+            } catch (IllegalArgumentException e) {
+                throw new InvalidReceiveException("Unknown API key " + apiKey);
+            }
+            if (apiVersion < 0) {
+                throw new InvalidReceiveException("Invalid api version " + apiVersion);
+            } else if (api.latestVersion() < apiVersion) {
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java:159]: Should we also throw InvalidReceiveException if apiVersion is less than oldestVersion?

File: clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java [hunk 2]
```diff
@@ -26,35 +28,165 @@
...
+            }
+            return read;
+        }
+    }
+
+    /**
+     * A header receiver which reads a Kafka Request.
+     */
+    public static class RequestStartHeaderReceiver implements HeaderReceiver {
+        private final static int MIN_REQUEST_SIZE = 10;
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java:112]: This is the request header size, right? Perhaps we could put the constant in RequestHeader and reference it here?

File: clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java [hunk 3]
```diff
@@ -26,35 +28,165 @@
...
+        private final int maxSize;
+        private final ByteBuffer startBuffer;
+        private int dataSize = -1;
+        private short apiKey = (short) -1;
+        private short apiVersion = (short) -1;
+
+        public RequestStartHeaderReceiver(int maxSize) {
+            this.maxSize = maxSize;
+            this.startBuffer = ByteBuffer.allocate(8);
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java:121]: Could we add a comment on what the 8 bytes are for?

**Review by junrao — COMMENTED** [review]
@cmccabe : Thanks for the explanation. Made a pass and left a few minor comments. Also, we explicitly handle ApiVersion and SaslHandshakeRequest in SaslServerAuthenticator. I am wondering if we should do the same header check there, which is probably trickier.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5731/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5745/
Test FAILed (JDK 7 and Scala 2.11).

**Pull request closed without merging**